### PR TITLE
Add StatefulSet support in OpenTelemetry Operator

### PR DESCRIPTION
**Pull request by IrisTuntun**
Add StatefulSet support in Operator so that the Operator can start collectors in StatefulSet mode.  Also added unit tests and e2e tests. Fixes  [wg-prometheus#31](https://github.com/open-telemetry/wg-prometheus/issues/31) and [wg-prometheus#47](https://github.com/open-telemetry/wg-prometheus/issues/47).

### CODE REVIEW

File: api/v1alpha1/opentelemetrycollector_types.go [hunk 1]
```diff
@@ -51,6 +51,12 @@ type OpenTelemetryCollectorSpec struct {
 	// +operator-sdk:gen-csv:customresourcedefinitions.specDescriptors=true
 	ServiceAccount string `json:"serviceAccount,omitempty"`
 
+	// VolumeClaimTemplates will provide stable storage using PersistentVolumes.
```
Comment (jpkrohling) [inline api/v1alpha1/opentelemetrycollector_types.go:54]:
```suggestion
	// VolumeClaimTemplates will provide stable storage using PersistentVolumes. Only available when the mode=statefulset.
```

File: pkg/collector/reconcile/statefulset.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+		// it exists already, merge the two if the end result isn't identical to the existing one
+		updated := existing.DeepCopy()
+		if updated.Annotations == nil {
+			updated.Annotations = map[string]string{}
+		}
+		if updated.Labels == nil {
+			updated.Labels = map[string]string{}
+		}
+
+		updated.Spec = desired.Spec
```
Comment (jpkrohling) [inline pkg/collector/reconcile/statefulset.go:83]: Could you double-check that you can copy the entire Spec here? Some resources have properties that can't be changed, or that we don't want to change.

File: pkg/collector/volumeclaim.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/open-telemetry/opentelemetry-operator/api/v1alpha1"
+	"github.com/open-telemetry/opentelemetry-operator/internal/config"
+)
+
+// VolumeClaimTemplates builds the volumeClaimTemplates for the given instance,
+// including the config map volume mount.
+func VolumeClaimTemplates(cfg config.Config, otelcol v1alpha1.OpenTelemetryCollector) []corev1.PersistentVolumeClaim {
```
Comment (jpkrohling) [inline pkg/collector/volumeclaim.go:29]: Question: given that this is only used when the mode=statefulset, should this function here be checking that condition and return an empty slice when that condition is not satisfied?

File: tests/e2e/smoke-statefulset/00-assert.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: stateful-collector
+status:
+  replicas: 1
```
Comment (jpkrohling) [inline tests/e2e/smoke-statefulset/00-assert.yaml:6]: Do Statefulsets have `readyReplicas` as well?

**Review by jpkrohling — COMMENTED** [review]
This looks great! The design doc was also very nice, thank you for that. I left a few comments, but nothing major. I'd be more than happy to review your future docs in the future before they become PRs. As I'm not part of the Prometheus WG, feel free to ping me directly.

**Comment by jpkrohling** [comment]
One extra comment: we have a validating webhook that can be used to ensure that the VolumeClaimTemplate is only used when the mode is statefulset:

https://github.com/open-telemetry/opentelemetry-operator/blob/b6ac997b1ba13d756f8679870d49ee497b699c6f/api/v1alpha1/opentelemetrycollector_webhook.go#L57-L67

File: pkg/collector/reconcile/statefulset.go (hunk 2 above)
Comment (IrisTuntun) [reply pkg/collector/reconcile/statefulset.go:83]: I have double checked [StatefulSetSpec](https://github.com/kubernetes/api/blob/3a2d6b5bb7f84daa2125131865eb81b151448d5b/apps/v1beta1/types.go#L162), and didn't see any fields that we do not want to change. Any ideas?

File: api/v1alpha1/opentelemetrycollector_types.go (hunk 1 above)
Comment (humivo) [reply api/v1alpha1/opentelemetrycollector_types.go]: Applied in latest commit: "Fix volumeclaim function and small bugs"

File: pkg/collector/volumeclaim.go (hunk 3 above)
Comment (humivo) [reply pkg/collector/volumeclaim.go:29]: Applied this check to the function in commit: "Fix volumeclaim function and small bugs"

File: tests/e2e/smoke-statefulset/00-assert.yaml (hunk 4 above)
Comment (IrisTuntun) [reply tests/e2e/smoke-statefulset/00-assert.yaml:6]: `readyReplicas` assertions added

**Comment by humivo** [comment]
@jpkrohling Also added the logic for ensuring volumeClaimTemplates only used for statefulset modes in the validation webhook

**Review by jpkrohling — APPROVED** [review]
Thank you for this PR!

**Pull request merged**